### PR TITLE
iterate non-enumerables in old IE also

### LINKED ISSUE #2613: ie8: non-iterable properties valueOf, constructor. Patch Object.forEach
In IE8 the constructor field if present and is a function is masked when iterating with for..in :
```
var x={a:7,constructor:function(){return 9;}}
'constructor' in x
true
for(var key in x)alert(key);
//only "a" pops out
```
This causes a bug in Object.each, in Object.merge, in Options.setOptions and so on.
This is particularly problematic for Neuro.js which insists on calling one of the options passed to initialize collection "constructor", as in this case constructor is usually a function (to be specific: a mootools Class extending Neuro.Model) which results in Neuro.js not working at all under ie8.

<bountysource-plugin>

---
Want to back this issue? **[Place a bounty on it!](https://www.bountysource.com/issues/2637375-object-each-setoptions-etc-fail-to-iterate-over-constructor-field-in-ie8?utm_campaign=plugin&utm_content=tracker%2F22067&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F22067&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by SergioCrisostomo**
fixes #2613

### CODE REVIEW

**Comment by megawac** [comment]
:+1: but why can't you guys implement all these methods by way of `Object.keys` which will also be faster in browsers that support `keys` natively

**Comment by SergioCrisostomo** [comment]
@megawac this was the best I came up with. Input is really welcome, to make it even better :)

In this case since its a IE8- fix i think `Object.keys` is not a option in IE8. (do correct me if I understood you wrong)

**Comment by megawac** [comment]
I mean `Mootools` polyfills `Object.keys` that also has this fix no?

**Comment by SergioCrisostomo** [comment]
@megawac oh, I see. Actually  [`Object.keys` is also depending on a `for var in`](https://github.com/mootools/mootools-core/blob/master/Source/Types/Object.js#L63-L69). Maybe it should use a patched `Object.each` instead...

**Comment by megawac** [comment]
@SergioCrisostomo I would say it should be the other way around (thats how most other frameworks otu there are doing it). Having `keys` be the root fix for the non enumerable bug in ie also lets other browsers benefit from a faster implementation of all the object methods as native `keys` is faster than `for .. in`. I would argue all the object methods (`each`, `map`, `filter`, etc) that iterate own properties should be using `keys`

**Comment by arian** [comment]
I think that'd be better indeed.

We have to move Object.keys to Core.js from Types/Object.js. After that we could rewrite all `for in` occurrences to use Object.keys.

**Comment by SergioCrisostomo** [comment]
Updated!

File: Source/Core/Core.js [hunk 1]
```diff
@@ -301,13 +306,24 @@ Number.extend('random', function(min, max){
...
-Object.extend('forEach', function(object, fn, bind){
-	for (var key in object){
-		if (hasOwnProperty.call(object, key)) fn.call(bind, object[key], key, object);
+Object.extend({
+
+	keys: function(object){
+		var keys = [];
+		eachKey(object, function(key, value){
+			if (hasOwnProperty.call(object, key)) keys.push(key);
+		}, this);
```
Comment (arian) [inline Source/Core/Core.js]: You don't need this `this` here..

File: Source/Core/Core.js [hunk 2]
```diff
@@ -301,13 +306,24 @@ Number.extend('random', function(min, max){
...
+		return keys;
+	},
+
+	forEach: function(object, fn, bind){
+		Object.keys(object).forEach(function(key){
```
Comment (arian) [inline Source/Core/Core.js]: Array.forEach is not defined at this point yet.. (but only a few lines after this). Maybe do the Object.* methods after the Array ones?

File: Source/Types/Object.js [hunk 3]
```diff
@@ -31,49 +31,46 @@ Object.extend({
...
-		return keys;
-	},
-
 	values: function(object){
-		var values = [];
-		for (var key in object){
-			if (hasOwnProperty.call(object, key)) values.push(object[key]);
-		}
-		return values;
+		return Object.keys(object).map(function(key){
```
Comment (arian) [inline Source/Types/Object.js]: Here you depend on one of the `Array` methods, which isn't required in the yaml. This code is definitely nicer, but not sure if we want to require Array for this, or just do the for-loop.

File: Specs/Core/Core.js [hunk 4]
```diff
@@ -790,6 +811,32 @@ describe('Object.each', function(){
 		expect(daysObj).toEqual({first: 'Sunday', second: 'Monday', third: 'Tuesday'});
 	});
 
+	/*<ltIE8>*/
```
Comment (arian) [inline Specs/Core/Core.js]: This is not just for lt IE8.

**Comment by SergioCrisostomo** [comment]
@arian updated again.

File: Source/Core/Core.js (hunk 2 above)
Comment (arian) [inline Source/Core/Core.js]: Actually, you can use `eachKey` here directly:

```js
forEach: eachKey
```

File: Source/Core/Core.js [hunk 5]
```diff
@@ -57,27 +57,32 @@ var instanceOf = this.instanceOf = function(item, object){
...
-
-var Function = this.Function;
-
+/*<ltIE8>*/
 var enumerables = true;
 for (var i in {toString: 1}) enumerables = null;
 if (enumerables) enumerables = ['hasOwnProperty', 'valueOf', 'isPrototypeOf', 'propertyIsEnumerable', 'toLocaleString', 'toString', 'constructor'];
+/*</ltIE8>*/
+var hasOwnProperty = Object.prototype.hasOwnProperty;
+function eachKey(object, fn, thisArg){
```
Comment (arian) [inline Source/Core/Core.js]: I think `eachKey` isn't a very nice name. maybe `objectEach`, or just `each`?

Also I'm thinking if this shouldn't be:

```js
var objectKeys = Object.keys || function(object){
   // ...
   return keys;
};
```

And use that in the overloadSetter, with a for-loop there.

And below, just have:

```js
Object.extend({
  keys: objectKeys,
  forEach: // ...
});
```

File: Source/Core/Core.js [hunk 6]
```diff
@@ -57,24 +57,38 @@ var instanceOf = this.instanceOf = function(item, object){
...
 var enumerables = true;
 for (var i in {toString: 1}) enumerables = null;
 if (enumerables) enumerables = ['hasOwnProperty', 'valueOf', 'isPrototypeOf', 'propertyIsEnumerable', 'toLocaleString', 'toString', 'constructor'];
+/*</ltIE8>*/
+var hasOwnProperty = Object.prototype.hasOwnProperty;
+function objectKeys(object, alsoPrototypeChain){
+	var keys = [];
+    for (var k in object){
+		if (alsoPrototypeChain) keys.push(k);
+		else hasOwnProperty.call(object, k) && keys.push(k);
```
Comment (arian) [inline Source/Core/Core.js]: Seriously, use `else if`, not the `x && y` thing!

**Comment by SergioCrisostomo** [comment]
rebased and ready...

**Comment by bhoomi01** [comment]
Thank you for sharing this very nice post awesome keep sharing.
[getapk market descargar](https://getapkmarket.site/download/)

**Comment by joanmelda** [comment]
[Economics Assignment Writing Services](https://researchpapers247.com/economics-assignment-help/) has become a significant challenge owing to the complexity of these [Economics Research Paper Writing Services](https://researchpapers247.com/economics-assignment-help/) and in other cases lack of time and knowledge needed to complete [Economics Coursework Writing Services](https://researchpapers247.com/economics-assignment-help/).

**Comment by documentwizard100** [comment]
anchorage dispensary
weed for sale
buy real weed online cheap
weed for sale online
weed delivery dc
420 dispensary
dank legal buds for sale
http://buymarijuanas.com

**Comment by documentwizard100** [comment]
anchorage dispensary
[weed for sale
](buymarijuanas.com)buy real weed online cheap
weed for sale online
weed delivery dc
420 dispensary
[dank legal buds for sale
](buymarijuanas.com)http://buymarijuanas.com

**Comment by Nortech001** [comment]
Nice post. I used to be checking constantly this blog and I am impressed! Very helpful information specially the last part I maintain such info much. I was seeking this particular info for a very lengthy time. Thank you and good luck.
https://www.mygeekshelp.com/macbook-stuck-on-apple-logo-screen/
macbook stuck on apple logo
macbook pro stuck on startup screen
macbook pro stuck on apple loading screen
macbook stuck at apple logo

**Pull request merged**